### ISSUE TITLE
Selenium Ruby (Grid), has a huge delay after calling driver.quit on IE11 when encountering Net::ReadTimeout

### ISSUE THREAD

**Issue by luke-hill** [SeleniumHQ/selenium#3043]
## Meta -
OS: Windows 7 / 8.1
Selenium Version:  At least 10 different ruby bindings from `2.53.4` through to `3.5.2`
Browser: IE11 on Grid

## Preface to our usage -

We increase our timeouts in the bindings to `120s` using the following syntax

```
Capybara.register_driver :selenium do |app|
    client = Selenium::WebDriver::Remote::Http::Default.new
    client.timeout = 120
    Capybara::Selenium::Driver.new(
      app,
      browser: :remote,
      url: ENV['hub_url'],
      desired_capabilities: caps,
      http_client: client,
    )
  end
```

## Expected Behavior -
When we timeout we expect the timestamps between running the last command and then receiving a timeout to be 120s +/- 0.25s (Based off the default interrogation / polling rate). This *works*

Then we run our after hook. Which is contained in the code below (Steps to reproduce). This should run in a couple of seconds.

## Actual Behavior -
This however has a "huge" delay between the first and second log statements. Circa **150s**

## Steps to reproduce -
```
After do |scenario|
  $LOG.info("Attempting to quit session.driver for #{scenario.name}") #This log is the penultimate
  Capybara.current_session.driver.quit
  $LOG.info('Quit driver for this scenario') #This log is the last and it takes 150s
end
```

**Comment by luke-hill** [comment]
For reference (Here is our last log so you can see the timestamp issues. I have raised this on the Capybara project and they advised me to come here: my problem here is with the **last 2 lines**

```
2016-10-26 09:57:54 +0100:INFO - Added credit card details
2016-10-26 09:57:54 +0100:INFO - Attempted to pay full amount
2016-10-26 09:59:56 +0100:INFO - FAILED: Flight & Hotel to Gran Canaria from Stockholm with full credit card payment ## This takes 2mins as we set our timeout to 2mins, we're happy with this
2016-10-26 09:59:56 +0100:INFO - REASON: Net::ReadTimeout
2016-10-26 09:59:56 +0100:INFO - DURATION: 00:03:20 secs.
2016-10-26 09:59:56 +0100:INFO - Attempting to quit session.driver for Flight & Hotel to Gran Canaria from Stockholm with full credit card payment
2016-10-26 10:02:25 +0100:INFO - Quit driver for this scenario ## This should occur 1-2 seconds after the above line
```

**Comment by luke-hill** [comment]
As an update. We have seen this happen on Firefox also - This delay took four minutes, and didn't even run the last verbose Command. The code is the same as above.

```
2016-11-18 10:59:02 +0000:INFO - FAILED: Flight to Tenerife from Stockholm for two adults with full payment
2016-11-18 10:59:02 +0000:INFO - REASON: Net::ReadTimeout
2016-11-18 10:59:02 +0000:INFO - DURATION: 02:45.
2016-11-18 10:59:02 +0000:INFO - Attempting to quit session.driver for Flight to Tenerife from Stockholm for two adults with full payment
2016-11-18 11:03:02 +0000:INFO - 
************************************************************************************************************************
```
The *'s is the next run starting.

**Comment by titusfortner** [comment]
I don't typically run the Ruby specs on the grid, but perhaps I should
start. Is the Firefox issue also on Windows?

**Comment by luke-hill** [comment]
All our machines are Windows Machines. Either running Windows 7 or Windows 8.1.

We have a variety of different IE/Win versions. Do you want the full list?

**Comment by luke-hill** [comment]
Something to add to this (I only found this out recently, as it's a hard error to manually generate). Is that this error *does not* occur on local machines. It is a grid only issue.

I managed to get a Net::ReadTimeout error on my local machine. below is the log output

```
2016-11-22 09:39:29 +0000:INFO - Attempted to pay full amount
2016-11-22 09:40:32 +0000:INFO - FAILED: Flight & Hotel to Gran Canaria from Stockholm with full credit card payment
2016-11-22 09:40:32 +0000:INFO - REASON: Net::ReadTimeout
2016-11-22 09:40:32 +0000:INFO - DURATION: 00:02:04 secs.
2016-11-22 09:40:32 +0000:INFO - Attempting to quit session.driver for Flight & Hotel to Gran Canaria from Stockholm with full credit card payment
2016-11-22 09:40:32 +0000:INFO - Quit driver for this scenario
2016-11-22 09:40:32 +0000:INFO - 
************************************************************************************************************************
2016-11-22 09:40:32 +0000:INFO - Starting feature: Flight and Hotel search
```

**Comment by titusfortner** [comment]
If you run this with `$DEBUG = true` you can see the output of what is sent to and received from the grid. This info could be useful to debugging.

**Comment by luke-hill** [comment]
I've had a little experiment with this command and the output is colossal. What I'll probably do is insert it in the `After` hook just before the command which crashes.

One issue is I have to run this locally, as the $DEBUG seems to output to my console, and not a specific `.log` file anywhere. I'll give it a full run tomorrow and see what we get from it, I'm not a Ruby expert though so it might be a bit cluttered.

**Comment by titusfortner** [comment]
Yeah, it goes to stdout. Revamping the Ruby logging is on my list of things to do. :)

**Comment by luke-hill** [comment]
I've not forgotten this, but also I've not got very far with this either. The output is plentiful, but I'm experimenting with it in different places to give you some meaningful output. I think the best bet is to only change the Debug state once im in After hooks, as otherwise it's just too big a log to send you

**Comment by luke-hill** [comment]
@titusfortner What I've found so far.

- The lag is *always* 150s. So it sounds to me maybe some value of 150s is being hardcoded somewhere.

- A positive response for `$DEBUG=true` (I.e. the test has passed), yields the following. (I only turned on DEBUG for the final few items) - On Internet Explorer grid (Locally ran but on remote grid)

```
-> DELETE session/3a351047-60d8-4b06-8886-515c9739151f
Exception `IO::EWOULDBLOCKWaitReadable' at C:/Ruby22/lib/ruby/2.2.0/net/protocol.rb:153 - A non-blocking socket operation could not be completed immediately. - read would block
<- {"state":"success","sessionId":"3a351047-60d8-4b06-8886-515c9739151f","hCode":26161955,"value":null,"class":"org.openqa.selenium.remote.Response","status":0}

Exception `NoMethodError' at C:/Ruby22/lib/ruby/2.2.0/delegate.rb:91 - undefined method `to_ary' for #<Cucumber::RunningTestCase::Scenario:0x4d63988>
Exception `NoMethodError' at C:/Ruby22/lib/ruby/2.2.0/delegate.rb:91 - undefined method `to_ary' for #<Cucumber::RunningTestCase::Scenario:0x4d30e80>
```

**Comment by luke-hill** [comment]
@titusfortner Big update on this.

Since updating to Selenium3, it is now taking 8 minutes to quit after a Net::Read::Timeout

See the following logs

```
2017-02-15 07:25:32 +0000:INFO - Attempted to add credit card details
2017-02-15 07:27:35 +0000:INFO - FAILED: User books a Hotel in Tenerife with full payment
2017-02-15 07:27:35 +0000:INFO - REASON: Net::ReadTimeout
2017-02-15 07:35:35 +0000:INFO - 
************************************************************************************************************************
2017-02-15 07:35:35 +0000:INFO - Starting feature: Place an order for Hotel Only, scenario: User books a Hotel in Gran Canaria with full payment
2017-02-15 07:35:35 +0000:INFO - With scenario tags: ["@e2e", "@hotel_only", "@email", "@se"]
2017-02-15 07:35:35 +0000:INFO - ************************************************************
2017-02-15 07:39:35 +0000:INFO - FAILED: User books a Hotel in Gran Canaria with full payment
2017-02-15 07:39:35 +0000:INFO - REASON: Net::ReadTimeout
2017-02-15 07:47:36 +0000:INFO - 
************************************************************************************************************************
```

Tagging @p0deje In this as he was working with the `$DEBUG` item recently. Is it worth maybe us switching to that ASAP to try diagnose this issue further. As this is something which is causing our runs to skyrocket on IE only.

**Comment by titusfortner** [comment]
That's not enough of a stack trace or $DEBUG output to see what is going on. Those errors in your 12/13 comment are normal and wouldn't cause an issue.

All the new Logger work will do is allow you to pipe to a file rather than STDOUT.

**Comment by luke-hill** [comment]
As of Selenium 3.2.2 we can now pipe this to a file. I have wrapped the problematic line (Basically pay for the holiday), including the line that causes the crash `button.click` in 2 open and close the selenium logger cukes.

The relating feature files all look like this

```
.......
And I start the Selenium Logger
And I add valid payment details
And I make a full payment <- this is the cuke that crashes
And I stop the Selenium Logger
Then I should see the order details
```

I've got a couple of legit outputs and a couple of crashed outputs all in diff files. What's the best way to upload these?

**Comment by lmtierney** [comment]
You can attach files by dragging and dropping, etc. Look at the bottom of the comment input for direction

**Comment by luke-hill** [comment]
[sel logs.zip](https://github.com/SeleniumHQ/selenium/files/824065/sel.logs.zip)

**Comment by luke-hill** [comment]
[selenium_node_172.23.4.115_debug.zip](https://github.com/SeleniumHQ/selenium/files/849944/selenium_node_172.23.4.115_debug.zip)

Attached a much cutdown log file. I've added in comments with `#######` the entire file is less than 800 lines so shouldn't be too tricky. I've put in a little bit before / after incase you needed that.

**Comment by shs96c** [comment]
After some debugging on IRC: this looks like the click is taking an unbelievably long time to finish, and everything is timing out because of this. What we really need is the log from the IEDriverServer, but I'm unsure how to persuade the node to return those to you.

**Comment by luke-hill** [comment]
New Driver level trace log now available
[iedriverserver_trace_log.txt](https://github.com/SeleniumHQ/selenium/files/1242407/iedriverserver_trace_log.txt)

**Comment by luke-hill** [comment]
Updated issue with the following new information...
- Have found more instances of Firefox encountering this issue.
- Have found that on Retest of Selenium 3 (A variety of versions, but not the latest), the issue of the delay spiking from 150s to 400s is now gone. *Note the initial problem is not yet remedied in Selenium versions up to 3.2.2*
- Have found the IE Driver level stacktrace is almost identical to the ones in Selenium 2.53 (So not posting)

Versions of Selenium now tested (Ruby): 2.53.4 / 3.0.0.beta / 3.0.5 / 3.0.7 / 3.2.0 / 3.2.2

Next plan is to test the 3.3/3.4 versions (Note that the IEDriver version is now 3.3 (We plan on going up to 3.4 soon)

**Comment by titusfortner** [comment]
Is this an issue on the latest versions of Selenium + Drivers?

**Comment by luke-hill** [comment]
We're working our way up this slowly. We had issues on 3.4.1 (Now remedied as 3.4.1 I've been told by Alex needs 3.5 grid).

Since 14th Sep update I can confirm. 3.4.0 fully has this issue and 3.4.4 also has this issue. We'll be upgrading to 3.5 soon

**Comment by luke-hill** [comment]
Updated test candidates (Included `3.4.4` and `3.5.0`).

Issue still occurs in terms of magnitude (150s give or take delay).

Frequency of occurrence "feels" less. We're seeing this issue probably 10-20% less, although I don't have solid stats for this. Obviously with the Firefox occurrences we are using the FirefoxDriver (As we haven't managed to upgrade to gecko yet), and for IE we were using the relevant version of the driver that corresponded to the bindings (So 3.0/3.2/3.4/3.5)

**Comment by luke-hill** [comment]
I'm aware this is starting to look like a bad smell, But I've ran through our logs for the last month or so across about 10 different jobs. And for Firefox (Since moving to Marionette). Our Net::ReadTimeout rate is ridiculously low (Less than 0.1%).

I only managed to find one instance of it recently. And the delay after hitting the Net::ReadTimeout to eventual suite kill was only a couple of seconds.

Providing I don't get any instances in the next month of this. I'm going to assume that geckodriver has fixed this problem for us.

Furthermore, this leads me to believe, that actually this issue might be an issue with the Driver Executables. And *not* with any part of the Selenium Codebase.

**Comment by p0deje** [comment]
@luke-hill What about IE? If the issue is still there, we should keep the issue open. Otherwise, let's close it once you're sure it's not Selenium bug.

**Comment by luke-hill** [comment]
Yep still an issue. Ran on some later versions (Updated top part to be a lot more concise). Still causing problems sadly. I'll keep this issue open as long as there's a problem. But I'll amend the description / title next month if we get no more FF instances (And I'll assume FF is now fixed)

**Comment by luke-hill** [comment]
Update - Have still observed this issue as true (Sadly), Have not got any more version histories to attach (3.5 series is the latest as we're having other issues getting 3.5, and therefore higher versions stable).

Have removed Firefox from the title, as the issue is not present on Marionette (Furthermore, the rate at which these timeouts occur on Marionette is actually *lower* than on Chromedriver (So thats a plus).

Have also amended the title to hint at this timeout only occuring when the failure is due to a `Net::ReadTimeout`

**Comment by luke-hill** [comment]
Update: (Haven't got huge stats for this), but it feels as though the problem is about the same across Ruby 2.3 and 2.5 environments (Kind of as expected). We previously had the issue on Ruby 2.2 also, although we upgraded from the former shortly after me joining the company so I don't have much stats on this.

Haven't been able to test later versions of Selenium / IE due to issues upgrading (Apologies). So the description in the post is still the most up to date (a variety from 2.53.4 - 3.5.0)

Also in the last 90+ days we've not had a single instance of this where there has been a timeout of longer than 2-3 seconds (After hitting a `Net::ReadTimeout`).

This leads me almost categorically to believe this is now only an IE driver / Selenium Ruby / Grid issue (Not sure which one / combination thereof it is).

**Comment by luke-hill** [comment]
Apologies for the mass bump after a couple of months. But I have a new theory. And it might be that this issue isn't based in Selenium but I'm not sure.

I found this error triggering again. This time on Chrome (In responsive emulator mode)

```
2018-07-04 15:13:50 +0100:INFO - FAILED: ...
2018-07-04 15:13:50 +0100:DEBUG - REASON: Net::ReadTimeout
2018-07-04 15:13:50 +0100:DEBUG - Writing to screenshot ...
2018-07-04 15:16:52 +0100:ERROR - Order object: #<Order:0x
```

Interestingly enough. This is the method invoked. The issue seems to crop up after the test has crashed and when I invoke `Capybara.current_session` the first time.

```
def log_failure_and_capture_screenshot
  $LOG.info("FAILED: #{scenario_name}")
  $LOG.debug("REASON: #{exception_message}")
  $LOG.debug("Writing to screenshot #{screenshot_name(name_choice)}")
  Capybara.current_session.save_screenshot(screenshot_name(name_choice), full: true)
end
```

@twalpole - I know it's a longshot. But when The Automation has encountered a `Net::ReadTimeout` and has crashed after it's timeout (Set in the driver), is there any chance that `Capybara.current_session` method call might lag for a couple of minutes and just stay stuck....

**Comment by twalpole** [comment]
@luke-hill Capybara doesn't do anything around network timeouts, it just asks Selenium to do things.  However if you're calling `driver.quit` (which Capybara knows nothing about) and then taking a screenshot after, Selenium will try to contact the service that it's already shutdown (in quit) with a no longer valid session id and you'll be at the mercy of whatever the network timeouts are for that connection

**Comment by luke-hill** [comment]
Notes for people visiting this old "Closed as Unfixed" issue.

- This issue has not been triaged properly in over 1 year.
- The issue was *confirmed fixed* in geckodriver `0.21-0.23` (Delay was in the order of seconds, which is tolerable)
- The issue was *confirmed present* in IEDriverServer `2.53.1 - 3.4` (Delay was 150s for 2.53-3.0, 3.1/3.2 the delay spiked to 8minutes, then 3.3-3.4 it was back down to 150s). But a variety of factors may have affected this
- The IE11 tests on 3.5 weren't conclusive, and as such can't be read into, but it seems as though the "frequency" with which the tests threw this error in 2.53 - 3.4 on the whole didn't change much (Roughly 2-5% decrease), but it seems as though going from 3.4->3.5 decreased the frequency by around 10% (Again to re-iterate there wasn't enough data)
- Ruby / Standalone Versions being used were always non-optimal, and as such using higher rubies / standalones (Which will be enforced from Se4), may have improved the consistency.
- The Infrastructure under test was often saturated, and had months where data was often skewed due to external inteferences.
- The AUT was never ran under a pure Selenium environment (Always Capybara wrapper), however this shouldn't disuade from the fact it felt as though this was a driver issue (As Chrome never triggered this, and Firefox when triggered didn't have the massive delay). Further to reinforce this was IE11 was always up to date, whereas the FF/Chrome browsers often were not.
- A couple of different deep-dive tests as mentioned earlier, yielded info that "hinted" that IE11 still believed the "click" was ongoing. Although we don't know much more than that.
